src/main/java/com/hubspot/jinjava/objects/collections/SizeLimitingPyList.java

```diff
@@ -0,0 +1,90 @@
+package com.hubspot.jinjava.objects.collections;
+
+import com.hubspot.jinjava.interpret.IndexOutOfRangeException;
+import com.hubspot.jinjava.objects.PyWrapper;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+public class SizeLimitingPyList extends PyList implements PyWrapper {
+  private int maxSize;
+
+  private SizeLimitingPyList(List<Object> list) {
+    super(list);
+  }
+
+  public SizeLimitingPyList(List<Object> list, int maxSize) {
+    super(list);
+    if (maxSize <= 0) {
+      throw new IllegalArgumentException("maxSize must be >= 1");
+    }
+
+    this.maxSize = maxSize;
+    if (list.size() >= maxSize) {
+      throw createOutOfRangeException(list.size());
+    }
+  }
+
+  @Override
+  public boolean append(Object e) {
+    if (size() >= maxSize) {
+      throw createOutOfRangeException(size() + 1);
+    }
+    return super.append(e);
+  }
+
+  @Override
+  public void insert(int i, Object e) {
+    if (size() >= maxSize) {
+      throw createOutOfRangeException(size() + 1);
+    }
+    super.insert(i, e);
+  }
+
+  @Override
+  public boolean add(Object element) {
+    if (size() >= maxSize) {
+      throw createOutOfRangeException(size() + 1);
+    }
+    return super.add(element);
+  }
+
+  @Override
+  public void add(int index, Object element) {
+    if (size() >= maxSize) {
+      throw createOutOfRangeException(size() + 1);
+    }
+    super.add(index, element);
+  }
+
+  @Override
+  public boolean addAll(int index, Collection<?> elements) {
+    if (size() + elements.size() >= maxSize) {
```

Comment:
Integer overflow might be a concern. It should be safer to do `size() >= maxSize - elements.size()`

Reply:
I think we will run out of memory long before this if an array has 2.4B items in it!